### PR TITLE
Uglify sourcemaps support

### PR DESCRIPTION
**Pull request by DeMoorJasper**
* Sourcemap support for builds (supporting uglify)
* Added functions to search closest mapping for both original and generated mappings
* Tried to make it as fast as possible (Based on binary search) 🚀

### CODE REVIEW

File: src/transforms/uglify.js [hunk 1]
```diff
@@ -14,18 +15,36 @@ module.exports = async function(asset) {
     },
     compress: {
       drop_console: true
-    }
+    },
+    sourceMap: asset.options.sourceMaps
+      ? {
+          filename: asset.relativeName
+        }
```
Comment (devongovett) [inline src/transforms/uglify.js]: I think uglify has a way to pass an input source map to it so you don't have to manually map it like you did below. See https://github.com/mishoo/UglifyJS2#source-map-options. Looks like the `content` option?

Comment (DeMoorJasper) [reply src/transforms/uglify.js]: I've seen it but that would add an extra stringify and parse operation to the building of the new sourcemap, which is the largest performance problem with source-map 0.6.1

**Comment by sant123** [comment]
¿Any advance on this? Just asking 😝

**Comment by DeMoorJasper** [comment]
@sant123 as far as i know it's ready to go, just waiting on the approval and merge from Devon

**Comment by DeMoorJasper** [comment]
@devongovett this should be ready to go

File: src/transforms/uglify.js [hunk 2]
```diff
@@ -11,18 +13,39 @@ module.exports = async function(asset) {
...
+      asset.sourceMap = await new SourceMap().extendSourceMap(
+        asset.sourceMap,
+        result.map
+      );
+    } else {
+      asset.sourceMap = result.map;
+    }
+  }
+
+  // Log all warnings
```
Comment (devongovett) [inline src/transforms/uglify.js]: I'd rather not log all of these. Uglify logs a lot of warnings and they are pretty much useless. I removed this in master already.

File: src/transforms/uglify.js [hunk 3]
```diff
@@ -11,18 +13,39 @@ module.exports = async function(asset) {
...
   }
 
   let result = minify(code, options);
+
   if (result.error) {
     throw result.error;
   }
 
+  if (result.map) {
+    result.map = await new SourceMap().addMap(JSON.parse(result.map));
```
Comment (devongovett) [inline src/transforms/uglify.js]: I think it's possible to get the map without converting to JSON first. https://github.com/mishoo/UglifyJS2/blob/master/lib/sourcemap.js#L94

Might be faster - maybe you could even access the original mappings from there so you don't need to decode it.

**Pull request merged**